src/ZuzaluOracle.sol

```diff
@@ -70,47 +166,108 @@ contract ZuzaluOracle is Owned {
         });
     }
 
-    /// @notice
+    ///@notice Get the root and and the depth from the data structures that are passed as inputs
+    /// @dev The function accepts a storage pointer, not the actual data
+    /// @param roots The roots array
+    /// @param index The index of the root to get
+    /// @param depths The depths mapping
+    function _getRootAndDepth(uint256[] storage roots, uint256 index, mapping(uint256 => uint256) storage depths)
+        internal
+        view
+        returns (uint256, uint256)
+    {
+        uint256 root = roots[roots.length - index];
+        uint256 depth = depths[root];
+        return (root, depth);
+    }
+    /// @notice Verifies a Semaphore proof for a particular signal and group. The group is specified by the root and depth
+    /// which is not one of the groups defined in the contract. It returns true if the proof is valid, and false otherwise.
+    /// @param _root The root of the merkle tree
+    /// @param _depth The depth of the merkle tree
+    /// @param _nullifierHash The hash of the nullifier
+    /// @param _signal The signal to verify
+    /// @param _externalNullifier The external nullifier
+    /// @param _proof The proof to verify
+
```

Comment:
nit: stylewise, no space before comment and function, space after function before comment